test-cgroup-util: Check return values

Werkov opened the pull request:
Reported by Coverity and
Fixes CID#1587767

Werkov commented:
So, the added return values pointed out to some variance in testing environments, commit messages have details.
The only failing CI is now fuzzer timing out in Fedora (unrelated).

yuwata commented:
/packit build

yuwata commented:
Packit is in red:
```
/* test_proc */
Found cgroup2 on /sys/fs/cgroup/, full unified hierarchy
src/test/test-cgroup-util.c:221: Assertion failed: expected "cg_pidref_get_unit(&pid, &unit)" to succeed, but got error: No such device or address
```

yuwata commented:
semaphoreci:
```
/* test_proc */
src/test/test-cgroup-util.c:215: Assertion failed: expected "cg_pidref_get_unit(&pid, &unit)" to succeed, but got error: No such device or address
```

Werkov commented:
Thanks for looking this up, @yuwata. What is that version `257.918.gd98ead7270.0` referring to? (I can't find it neither in systemd/systemd repo nor systemd-team/systemd Debian repo). I'm curious whether the changes from last [force push](https://github.com/systemd/systemd/compare/0ae8a33656b9f76c79c4c4fa0bc6876c088aabf3..263989feec65093d89153ebaa994eb1dfcddf8bf) were rolled to that test env.

(If so, I'm secondarily wondering which process trips this, so I may add some more verbosity to debug this.)

Werkov commented:
What a coincidence with https://github.com/systemd/systemd/issues/36098, perhaps the test could be extended to cover capsules too.

yuwata commented:
This time, the test running on lxc in ubuntu CIs:
```
55376s /* test_proc */
55376s src/test/test-cgroup-util.c:225: Assertion failed: expected "cg_pidref_get_unit(&pid, &unit)" to succeed, but got error: No such device or address
(snip)
55472s autopkgtest [07:31:20]: test unit-tests: -----------------------]
55472s unit-tests           FAIL non-zero exit status 1
55472s autopkgtest [07:31:20]: test unit-tests:  - - - - - - - - - - results - - - - - - - - - -
55472s autopkgtest [07:31:20]: @@@@@@@@@@@@@@@@@@@@ summary
55472s boot-and-services    PASS
55472s unit-tests           FAIL non-zero exit status 1
55476s autopkgtest [07:31:24]: test tests-in-lxd: -----------------------]
55476s tests-in-lxd         FAIL non-zero exit status 4
55476s autopkgtest [07:31:24]: test tests-in-lxd:  - - - - - - - - - - results - - - - - - - - - -
```
Also, semaphore CI are red, though I am not familiar with its environment.

yuwata commented on src/basic/cgroup-util.c:
```diff
@@ -1295,23 +1298,28 @@ static const char *skip_user_manager(const char *p) {
         n = strcspn(p, "/");
-        if (n < STRLEN("user@x.service"))
+        if (n < STRLEN("user@x.service")) /* || n < STRLEN("capsule@x.service"))*/
+                return NULL;
+        unit_name = strndupa(p, n);
```
Not a matter, but please insert an empty line after `return NULL;`
Also, use `const char *unit_name`, if it will not be modified.

yuwata commented on src/basic/cgroup-util.c:
```diff
@@ -1295,23 +1298,28 @@ static const char *skip_user_manager(const char *p) {
-
-                memcpy(buf, p + 5, n - 5 - 8);
-                buf[n - 5 - 8] = 0;
+        /* Note that user manager services never need unescaping,
+         * since they cannot conflict with the kernel's own
+         * names, hence we don't need to call cg_unescape()
+         * here. */
```
When you touch the code, could you rebreak the comment with 109ch per line.

yuwata commented on src/basic/cgroup-util.c:
```diff
@@ -1295,23 +1298,28 @@ static const char *skip_user_manager(const char *p) {
 
-                if (parse_uid(buf, NULL) < 0)
+                return p;
+        } else if (strncmp(unit_name, "capsule@", 8) == 0) {
```
Also here:
```c
} else if (startswith(unit_name, "capsule@")) {
```

yuwata commented on src/basic/cgroup-util.c:
```diff
@@ -1287,6 +1287,9 @@ static const char *skip_session(const char *p) {
  * Skip user@*.service, but require it to be there.
  */
 static const char *skip_user_manager(const char *p) {
+        _cleanup_free_ char *i = NULL;
+        char *unit_name;
+        UnitNameFlags type;
```
Not a matter, but nowadays we allow to declare variables any places.
Please declare them when used.

yuwata reviewed (changes requested):
LGTM, but several coding style requests.

Werkov commented:
Wow, thatnks for thorough coding style review. Hopefully: a) it sticks in my head, b) won't slow down the path-string operations visibly.

poettering commented on src/test/test-cgroup-util.c:
```diff
@@ -210,14 +210,20 @@ TEST(proc) {
+
+                /* Not all processes belong to a specific user or a machine */
+                r = cg_pidref_get_owner_uid(&pid, &uid);
+                ASSERT_TRUE(r == 0 || r == -ENXIO);
```
doesn't matter much, but maybe do:

```c
ASSERT_TRUE(IN_SET(cg_pidref_get_owner_uid(…), 0, -ENXIO);
```

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1295,26 +1295,32 @@ static const char *skip_user_manager(const char *p) {
 
         n = strcspn(p, "/");
-        if (n < STRLEN("user@x.service"))
+        if (n < STRLEN("user@x.service")) /* || n < STRLEN("capsule@x.service")) */
```
maybe write as:

```c
if (n < CONST_MIN(STRLEN(…), STRLEN(…)))
        …
```

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1295,26 +1295,32 @@ static const char *skip_user_manager(const char *p) {
         p += strspn(p, "/");
```
please update comment above

poettering commented on src/basic/cgroup-util.c:
```diff
@@ -1295,26 +1295,32 @@ static const char *skip_user_manager(const char *p) {
-        if (memcmp(p, "user@", 5) == 0 && memcmp(p + n - 8, ".service", 8) == 0) {
-                char buf[n - 5 - 8 + 1];
+        _cleanup_free_ char *i = NULL;
+        const char *unit_name = strndupa_safe(p, n);
```
maybe don't use strndupa() here anymore. the input string might be attacker controlled, who knows... just use strndup(), htere's no shame in that

poettering replied:
or if you want to avoid the oom handling, then please add a check that validates that `n` is <= NAME_MAX before you do the strndupa() her

Werkov replied:
OK, I'll do dynamic allocation, user units may be deep hierarchies.

Werkov commented:
I addressed the feedback for capsule handling. I'll think more about the approach to test-cgroup-util and handling external processes.
I cherry-picked the capsule bugfix PR #36664 so that it isn't blocked on that test (I should've never touched that). When this PR is rebased, the bottom commits disappear.

yuwata commented:
Please rebase.

Werkov commented:
1st force push=rebase, 2nd force push=generalization of hidden cgroups.
